src/lib.rs

```diff
@@ -373,13 +402,13 @@ pub fn into_future(awaitable: &PyAny) -> PyResult<impl Future<Output = PyResult<
     let py = awaitable.py();
     let (tx, rx) = oneshot::channel();
 
-    let task = CREATE_TASK
-        .get()
-        .expect(EXPECT_INIT)
-        .call1(py, (awaitable, get_event_loop(py)))?;
-    let on_complete = PyTaskCompleter { tx: Some(tx) };
-
-    task.call_method1(py, "add_done_callback", (on_complete,))?;
+    CALL_SOON.get().expect(EXPECT_INIT).call1(
+        py,
+        (PyEnsureFuture {
+            awaitable: awaitable.into(),
+            tx: Some(tx),
+        },),
+    )?;
```

Comment:
How come the `ensure_future` call needs to be wrapped in `call_soon`? If it's to make sure that `ensure_future` is associated with the event loop, it might be possible to use the `loop` keyword parameter.

Reply:
So _I think_ that the way `ensure_future` schedules `coroutine` objects is not threadsafe. There are two functions for scheduling coroutines: `run_coroutine` and `run_coroutine_threadsafe`. We want to use the latter since we are likely scheduling it from another thread. According to [cpython](https://github.com/python/cpython/blob/755f3c1521b422bc2177013d289f5439975fdc4f/Lib/asyncio/tasks.py#L859), `run_coroutine_threadsafe` uses `call_soon_threadsafe` in order to schedule a call to `ensure_future` just like I'm doing here.

By reading up on this I also realized that we are using `event_loop.create_future` to create an `asyncio.Future` to propagate the result back to the event loop. It turns out `asyncio.Future` is not threadsafe, so we should probably change it to [`concurrent.futures.Future`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future)

Reply:
Ah I see, thanks for the explanation!

Reply:
Sorry being late to this but why dont we just use `asyncio.run_coroutine_threadsafe` which spawns the coroutine on the passed event loop in a threadsafe mannor and also gives us a usable handle back?

Reply:
Np, I didn't really make it clear that `asyncio.run_coroutine_threadsafe` was throwing the exception. Thanks for taking the time to review this @ChillFish8, I appreciate it!